api/middleware/stats.go

```diff
@@ -0,0 +1,86 @@
+package middleware
+
+import (
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/metrictank/stats"
+	"gopkg.in/macaron.v1"
+)
+
+type requestStats struct {
+	sync.Mutex
+	responseCounts    map[string]map[int]*stats.Counter32
+	latencyHistograms map[string]*stats.LatencyHistogram15s32
+	sizeMeters        map[string]*stats.Meter32
+}
+
+func (r *requestStats) PathStatusCount(path string, status int) {
+	r.Lock()
+	p, ok := r.responseCounts[path]
+	if !ok {
+		p = make(map[int]*stats.Counter32)
+		r.responseCounts[path] = p
+	}
+	c, ok := p[status]
+	if !ok {
+		c = stats.NewCounter32(fmt.Sprintf("api.request.%s.%d", pathSlug(path), status))
+		p[status] = c
+	}
+	r.Unlock()
+	c.Inc()
+}
+
+func (r *requestStats) PathLatency(path string, dur time.Duration) {
+	r.Lock()
+	p, ok := r.latencyHistograms[path]
+	if !ok {
+		p = stats.NewLatencyHistogram15s32(fmt.Sprintf("api.request.%s.latency", pathSlug(path)))
+		r.latencyHistograms[path] = p
+	}
+	r.Unlock()
+	p.Value(dur)
+}
+
+func (r *requestStats) PathSize(path string, size int) {
+	r.Lock()
+	p, ok := r.sizeMeters[path]
```

Comment:
i'm concerned about different, but equivalent paths ending up with different metrics. like `/node` and `/node/`. perhaps we should also use `pathSlug(path)` when looking up the metrics.